presto-hudi/src/main/java/com/facebook/presto/hudi/split/HudiBackgroundSplitLoader.java

```diff
@@ -0,0 +1,109 @@
+    private final ExtendedHiveMetastore metastore;
+    private final HudiTableLayoutHandle layout;
+    private final HoodieTableFileSystemView fsView;
+    private final AsyncQueue<ConnectorSplit> asyncQueue;
+    private final List<String> partitions;
+    private final String latestInstant;
+    private final int splitGeneratorNumThreads;
+    private final ExecutorService splitGeneratorExecutorService;
+
+    public HudiBackgroundSplitLoader(
+            ConnectorSession session,
+            ExtendedHiveMetastore metastore,
+            HudiTableLayoutHandle layout,
+            HoodieTableFileSystemView fsView,
+            AsyncQueue<ConnectorSplit> asyncQueue,
+            List<String> partitions,
+            String latestInstant)
+    {
+        this.session = requireNonNull(session, "session is null");
+        this.metastore = requireNonNull(metastore, "metastore is null");
+        this.layout = requireNonNull(layout, "layout is null");
+        this.fsView = requireNonNull(fsView, "fsView is null");
+        this.asyncQueue = requireNonNull(asyncQueue, "asyncQueue is null");
+        this.partitions = requireNonNull(partitions, "partitions is null");
+        this.latestInstant = requireNonNull(latestInstant, "latestInstant is null");
+
+        this.splitGeneratorNumThreads = getSplitGeneratorParallelism(session);
+        this.splitGeneratorExecutorService = Executors.newFixedThreadPool(splitGeneratorNumThreads);
+    }
+
+    @Override
+    public void run()
+    {
+        HoodieTimer timer = new HoodieTimer().startTimer();
+        List<HudiPartitionSplitGenerator> splitGeneratorList = new ArrayList<>();
+        List<Future> splitGeneratorFutures = new ArrayList<>();
+        ArrayDeque<String> partitionQueue = new ArrayDeque<>(partitions);
+
+        // Start a number of partition split generators to generate the splits in parallel
+        for (int i = 0; i < splitGeneratorNumThreads; i++) {
+            HudiPartitionSplitGenerator generator = new HudiPartitionSplitGenerator(
+                    session, metastore, layout, fsView, asyncQueue, partitionQueue, latestInstant);
+            splitGeneratorList.add(generator);
+            splitGeneratorFutures.add(splitGeneratorExecutorService.submit(generator));
+        }
+
+        for (HudiPartitionSplitGenerator generator : splitGeneratorList) {
+            // Let the split generator stop once the partition queue is empty
+            generator.stopRunning();
+        }
+
+        // Wait for all split generators to finish
+        for (Future future : splitGeneratorFutures) {
+            try {
+                future.get();
+            }
+            catch (InterruptedException | ExecutionException e) {
+                throw new HoodieException("Error generating Hudi split", e);
+            }
+        }
+        asyncQueue.finish();
+        LOG.debug(String.format("Finish getting all splits in %d ms", timer.endTimer()));
```

Comment:
Airlift logger supports format string. Let's try `log.debug("Finish ... %s ms", timer.endTimer())`

Reply:
Right.  I fixed all the occurrences.